src/lib/server/server_session.cpp

```diff
@@ -265,7 +265,8 @@ template <typename TConnection, typename TTaskRunner>
 boost::future<void> ServerSessionImpl<TConnection, TTaskRunner>::_handle_sync_command() {
   if (!_transaction) return boost::make_ready_future();
 
-  _transaction->commit();
+  const auto success = _transaction->commit();
+  Assert(success, "Commit has failed.");
```

Comment:
Transaction conflicts are, to some degree, expected in multi-threaded environments. Killing the server is thus not thecorrect answer. The server should respond with whatever psql does in this case. Also, this should be tested.

Reply:
Right now this is pretty difficult for me to test, because (at least until unique constraints are available) there is no way I can think of to provoke such an unsuccessful commit without having access to the actual transaction context.

Reply:
With the rare exception of `AssertInput`, we don't want to use exceptions as part of regular execution. `Fail` should always mean "Something is broken, we don't know why, and we go belly-up".

Can we pass the information to `_handle_client_requests` in a better way?

> Right now this is pretty difficult for me to test, because (at least until unique constraints are available) there is no way I can think of to provoke such an unsuccessful commit without having access to the actual transaction context.

Can't you just insert a hard-coded transaction failure?